### PR TITLE
Add RealmInteger data type, describe in-memory realms

### PR DESCRIPTION
**Pull request by MongoCaleb**
### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-14587
https://jira.mongodb.org/browse/DOCSP-14588

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/fruit/sdk/dotnet/examples/open-and-close-a-local-realm/#in-memory-realms
https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/fruit/sdk/dotnet/data-types/field-types/#realminteger-t-

### Review Guidelines
[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/dotnet/data-types/field-types.txt [hunk 1]
```diff
@@ -38,6 +39,54 @@ within {+client-database+}.
...
+When using a ``RealmInteger``, however, you can call the ``Increment()`` and 
+``Decrement()`` methods, and to reset the counter, you set it to ``0``, just as 
+you would an ``int``:
+
+.. literalinclude:: /examples/generated/dotnet/DataTypes.codeblock.realmint-use.cs
+   :language: csharp
+
+.. important::
+
+   When you reset a ``RealmInteger``, you may run into the offline merge issue
```
Comment (nirinchev) [inline source/sdk/dotnet/data-types/field-types.txt:83]: Perhaps we can clarify that by saying that all `Increment/Decrement` calls that happened chronologically before the reset will be discarded while the ones that happened after will be applied on the reset value.

File: source/sdk/dotnet/data-types/field-types.txt [hunk 2]
```diff
@@ -38,6 +39,54 @@ within {+client-database+}.
...
+
+.. literalinclude:: /examples/generated/dotnet/DataTypes.codeblock.realmint-use.cs
+   :language: csharp
+
+.. important::
+
+   When you reset a ``RealmInteger``, you may run into the offline merge issue 
+   described above.
+
+A ``RealmInteger`` is backed by traditional ``integer`` type, so no schema
```
Comment (nirinchev) [inline source/sdk/dotnet/data-types/field-types.txt:86]: Technically, no schema migration is required when migrating between all integer types - i.e. one can freely switch between `long`, `short`, `RealmInteger<int>` and so on. That being said, that might be a bit too complicated to explain, so leaving it as it is should be fine.

**Review by nirinchev — APPROVED** [review]
Looks good - just one suggestion about the conflict resolution note.

File: source/sdk/dotnet/data-types/field-types.txt (hunk 1 above)
Comment (MongoCaleb) [reply source/sdk/dotnet/data-types/field-types.txt:83]: @nirinchev I think this is obvious, no?

File: source/sdk/dotnet/examples/open-and-close-a-local-realm.txt [hunk 3]
```diff
@@ -28,10 +28,31 @@ configuration information:
 .. literalinclude:: /examples/generated/dotnet/Examples.codeblock.local-realm.cs
    :language: csharp
 
+In-Memory Realms
+----------------
+With an 
+:dotnet-sdk:`InMemoryConfiguration <reference/Realms.InMemoryConfiguration.html>` 
+object, you can create a {+realm+} that runs entirely in memory (without being 
+persisted to disk.) The following example shows how to do this:
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/examples/open-and-close-a-local-realm.txt:36]: Should we trim the parens here? I feel like they don't have an impact on the sentence flow.

File: source/sdk/dotnet/examples/open-and-close-a-local-realm.txt [hunk 4]
```diff
@@ -28,10 +28,31 @@ configuration information:
...
+object, you can create a {+realm+} that runs entirely in memory (without being 
+persisted to disk.) The following example shows how to do this: 
+
+.. literalinclude:: /examples/generated/dotnet/Program.codeblock.in-memory.cs
+   :language: csharp
+
+In-memory {+realm+}s might still use disk space if memory is running low, but all 
+files created by an in-memory {+realm+} will be deleted when the {+realm+} is 
+closed. When creating an in-memory {+realm+}, the identifier must be unique to all 
+{+realm+}s, including both in-memory and persisted {+realm+}s.
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/examples/open-and-close-a-local-realm.txt:44]: I should add this to the Android in-memory docs-- this is an interesting point I didn't include!

File: source/sdk/dotnet/data-types/field-types.txt (hunk 1 above)
Comment (nirinchev) [reply source/sdk/dotnet/data-types/field-types.txt:83]: Not necessarily - people may think that a client that reset the counter but has been offline for 3 months will just override any increments that have happened in the meantime, but that's not the case.

File: source/sdk/dotnet/data-types/field-types.txt [hunk 5]
```diff
@@ -38,6 +39,54 @@ within {+client-database+}.
 
 If you choose to provide custom constructors, you must declare a public constructor with no arguments.
 
+.. _realminteger:
+
+RealmInteger<T>
+---------------
+
+Realm offers 
+:dotnet-sdk:`RealmInteger<T> <reference/Realms.RealmInteger-1.html>` as a
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/data-types/field-types.txt:48]: this link is broken for me

File: source/sdk/dotnet/data-types/field-types.txt [hunk 6]
```diff
@@ -38,6 +39,54 @@ within {+client-database+}.
 
 If you choose to provide custom constructors, you must declare a public constructor with no arguments.
 
+.. _realminteger:
+
+RealmInteger<T>
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/data-types/field-types.txt:44]:
```suggestion
Realm Integer
```
I don't think we use literal code like this in titles anywhere else in the realm docs -- maybe just use the english version of the name instead?

File: source/sdk/dotnet/data-types/field-types.txt [hunk 7]
```diff
@@ -38,6 +39,54 @@ within {+client-database+}.
...
+---------------
+
+Realm offers 
+:dotnet-sdk:`RealmInteger<T> <reference/Realms.RealmInteger-1.html>` as a 
+special integer type to be used as a logical counter. ``RealmInteger<T>`` 
+exposes an additional API that can more clearly express intent and generate 
+better conflict resolution steps when using Synchronized Realms. The type 
+argument <T> can be of type ``byte``, ``short``, ``int``, or ``long``. The
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/data-types/field-types.txt:52]:
```suggestion
argument ``<T>`` can be of type ``byte``, ``short``, ``int``, or ``long``. The 
```
Should probably monospace literal code

File: source/sdk/dotnet/data-types/field-types.txt [hunk 8]
```diff
@@ -38,6 +39,54 @@ within {+client-database+}.
...
+exposes an additional API that can more clearly express intent and generate 
+better conflict resolution steps when using Synchronized Realms. The type 
+argument <T> can be of type ``byte``, ``short``, ``int``, or ``long``. The 
+following example shows how to use a ``RealmInteger`` property that maps to 
+an ``int``:
+
+.. literalinclude:: /examples/generated/dotnet/DataTypes.codeblock.realmint-class.cs
+   :language: csharp
+
+Traditionally, a counter would be implemented by reading a value, incrementing
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/data-types/field-types.txt:59]: passive -> active voice?
```suggestion
Traditionally, you might implement a counter by reading a value, incrementing 
```

**Review by nathan-contino-mongo — APPROVED** [review]
LGTM with some thoughts and suggestions

File: source/sdk/dotnet/examples/open-and-close-a-local-realm.txt (hunk 3 above)
Comment (MongoCaleb) [reply source/sdk/dotnet/examples/open-and-close-a-local-realm.txt:36]: I kept them, but re-worded to be a bit more clear.

**Pull request merged**